### ISSUE TITLE
ViewNotFoundException even know the views are in the build output

### ISSUE THREAD

**Issue by tidusjar** [NancyFx/Nancy#2330]
Exception: `Nancy.RequestExecutionException: Oh noes! ---> Nancy.ViewEngines.ViewNotFoundException: Unable to locate view 'Search/Index'`

The problem is that the output folder actually contains the standard `Views/Search/Index.cshtml`.

This does not happen when running locally (through Visual Studio). It happens when running it from the build output.

Any ideas?

**Comment by thecodejunkie** [comment]
Is this a self-hosted application?

**Comment by tidusjar** [comment]
Correct, You can see how I am starting the application here:
https://github.com/tidusjar/RequestPlex/blob/master/RequestPlex.UI/Program.cs

    using (WebApp.Start<Startup>(uri))
    {
        Console.WriteLine($"Request Plex is running on {uri}");
        Console.WriteLine("Press any key to exit");
        Console.ReadLine();
    }

**Comment by thecodejunkie** [comment]
Yeah but what Nancy hosting adapter are you using? I'm guessing it's something with the root path that's not functioning correctly. Are you using any of the `Nancy.Hosting.*` packages?

**Comment by khellang** [comment]
@tidusjar Since your module is called `SearchModule`, I suspect you could just do `return View["Index"]`? Can you post the list of locations Nancy tried to search?

**Comment by tidusjar** [comment]
@thecodejunkie I am not using any `Nancy.Hosting.*` packages, but I am using `Microsoft.Owin.Hosting`.

@khellang I am not 100% sure (New to Nancy).
My Views folder structure is the following:

    Views
    ---- Search
         ---- Index
    ---- Home etc

Here is the whole stacktrace:

```
Nancy.RequestExecutionException: Oh noes! ---> Nancy.ViewEngines.ViewNotFoundException: Unable to locate view 'Search/Index'
Currently available view engine extensions: sshtml,html,htm
Locations inspected: views/Search/Search/Index-en-US,views/Search/Search/Index,Search/Search/Index-en-US,Search/Search/Index,views/Search/Index-en-US,views/Search/Index,Search/Index-en-US,Search/Index
Root path: C:\Program Files\Request Plex\Release\
If you were expecting raw data back, make sure you set the 'Accept'-header of the request to correct format, for example 'application/json'
   at Nancy.ViewEngines.DefaultViewFactory.GetRenderedView(String viewName, Object model, ViewLocationContext viewLocationContext)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute4[T0,T1,T2,T3,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3)
   at Nancy.ViewEngines.DefaultViewFactory.RenderView(String viewName, Object model, ViewLocationContext viewLocationContext)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute4[T0,T1,T2,T3,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3)
   at Nancy.Responses.Negotiation.ViewProcessor.Process(MediaRange requestedMediaRange, Object model, NancyContext context)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute4[T0,T1,T2,T3,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2, T3 arg3)
   at Nancy.Responses.Negotiation.DefaultResponseNegotiator.NegotiateResponse(IEnumerable`1 compatibleHeaders, NegotiationContext negotiationContext, NancyContext context)
   at Nancy.Responses.Negotiation.DefaultResponseNegotiator.CreateResponse(IList`1 compatibleHeaders, NegotiationContext negotiationContext, NancyContext context)
   at Nancy.Responses.Negotiation.DefaultResponseNegotiator.NegotiateResponse(Object routeResult, NancyContext context)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)
   at Nancy.Routing.DefaultRouteInvoker.<>c__DisplayClassa.b__6(Task`1 completedTask)
   --- End of inner exception stack trace ---
   at Nancy.NancyEngine.InvokeOnErrorHook(NancyContext context, ErrorPipeline pipeline, Exception ex)
```

**Comment by grumpydev** [comment]
You have something whacky going on because it's looking in Search/Search - can you post the module?

**Comment by grumpydev** [comment]
You haven't installed the razor view engine package.

**Comment by grumpydev** [comment]
Currently available view engine extensions: sshtml,html,htm

^^ that's just the built in SSVE view engine, you need to install Nancy.ViewEngines.Razor in order to use Razor (cshtml) views.

**Comment by thecodejunkie** [comment]
He hast he package installed, might be a case of the .net linker removing the reference (deferred loading)

try putting

`var foo = typeof(RazorViewEngine)` anywhere in your code, before you actual start up Nancy. That will give us an indication if that's the problem

**Comment by tidusjar** [comment]
@grumpydev Strange, If I copy my build output manually over to the server it all works. But If I use the build output from the CI build it doesn't (`Nancy.ViewEngines.Razor.dll` is present there).

**Comment by grumpydev** [comment]
Anything to do with requireReinstallation="true" being all over the packages. file?

**Comment by grumpydev** [comment]
Is the CI build copying the MS Razor DLL too? Sounds like something is trying to be "clever" in what it deploys depending on what it thinks you need :)

**Comment by tidusjar** [comment]
Yeah... You can see what it copies over here: https://ci.appveyor.com/project/tidusjar/requestplex

**Comment by tidusjar** [comment]
Found the problem, I'm hoping you guys can explain. Inside my bin (output) folder I have the following
    Debug (folder)
    Release (folder)
    Nancy.ViewEngines.Razor.BuildProviders.dll
    Nancy.ViewEngines.Razor.dll

I copied the two dll's above onto the server and replaced what was existing and it now works.

**Comment by tidusjar** [comment]
Found it. The `Nancy.ViewEngines.Razor.dll` was 'blocked' I had to right click -> Properties -> unblock. It now works... I wonder why windows was blocking it.

Thank you for the assistance anyway :+1:

**Comment by grumpydev** [comment]
Ah, it will block it if you have a ZIP file you've downloaded from the internet, and if you extract without unblocking the ZIP it will block all the DLLs which then .net won't load :)

Interesting project btw :)

**Comment by tidusjar** [comment]
Ah that's why! 
Thanks! It should be a replacement for the Original Plex Requests. I'm quite new to Nancy so I'm probably not doing things 100% correctly. I will be revisiting the documentation soon :)

**Comment by thecodejunkie** [comment]
@tidusjar one thing you can change in your modules is call `:base()` on them. It will let you pass in a `module path` and all routes, which are declared in the same module` will be relative to that. So no need to do

```c#
Get["/search"]
Get["/search/foo"]
Get["/search/bar"]
```

they will simply be

```c#
Get["/"]
Get["/foo"]
Get["/bar"]
```

https://github.com/NancyFx/Nancy/wiki/Exploring-the-nancy-module#using-modules-to-create-a-root-for-your-routes

**Comment by phillip-haydon** [comment]
What are your static files? All HTML files?

HTML files are considered Views so you could do a directory browse to create an in-memory list of files in the folder. Then do a single route like:

```
Get["/statics/{file}"]...
```

This uses the `_.file` value to look up if its a file from the directory (so that users don't attempt to navigate to files by passing in `../web.config` and if it's ok then just doing

`return View[(string)_.file];`

**Comment by tidusjar** [comment]
@thecodejunkie Thanks! I discovered that yesterday, handy!

@phillip-haydon My static files are `.cshtml`,`.js` and `.css`. Do you have some sort of example I could look at?

**Comment by phillip-haydon** [comment]
`.cshtml` or `.html`? If they are razor files I would assume you want to process them before returning them...

For `.js` and `.css` can't they just come from `./content` folder which is already a folder that nancy ignores and serves as static content?

**Comment by tidusjar** [comment]
@phillip-haydon I'm using `.cshtml` so they are all razor files. And yes all of my `.js` and `.css` are in the `/content` folder. 

So what are you suggesting? I'm a bit lost now.

**Comment by phillip-haydon** [comment]
Sorry @tidusjar I'm getting confused between issues being raised. Mixed this ticket with another one just raised an hour ago that I replied to.
